Ak/main readme (#71)
huggingface/optimum-intel: README.md

Intel [Neural Compressor](https://www.intel.com/content/www/us/en/developer/tools/oneapi/neural-compressor.html) is an open-source library enabling the usage of the most popular compression techniques such as quantization, pruning and knowledge distillation. It supports automatic accuracy-driven tuning strategies in order for users to easily generate quantized model. The users can easily apply static, dynamic and aware-training quantization approaches while giving an expected accuracy criteria. It also supports different weight pruning techniques enabling the creation of pruned model giving a predefined sparsity target.

> [OpenVINO](https://docs.openvino.ai/latest/index.html) is an open-source toolkit that enalbes high performance inference capabilities for Intel CPUs, GPUs, and special DL inference accelerators. It also provides a set of tools to optimize and quantize models. Optimum Intel provides a simple interface to convert a model to OpenVINO IR format and to run inference using OpenVINO.

Reviewer: ```suggestion
[OpenVINO](https://docs.openvino.ai/latest/index.html) is an open-source toolkit that enables high performance inference capabilities for Intel CPUs, GPUs, and special DL inference accelerators. It also provides a set of tools to optimize and quantize models. Optimum Intel provides a simple interface to convert a model to OpenVINO IR format and to run inference using OpenVINO.
```

To install the latest release of this package with the corresponding required dependencies, you can do respectively:
| Accelerator                                                                                                      | Installation                                      |
|:-----------------------------------------------------------------------------------------------------------------|:--------------------------------------------------|
| [Intel Neural Compressor](https://www.intel.com/content/www/us/en/developer/tools/oneapi/neural-compressor.html) | `python -m pip install optimum[neural-compressor]`|

> | [OpenVINO](https://docs.openvino.ai/latest/index.html)                                                           | `python -m pip install optimum[openvino,nncf]`         |

Reviewer 2: Yes you are right ! We will be able to change it back after optimum next release

> | [OpenVINO](https://docs.openvino.ai/latest/index.html)                                                           | `python -m pip install optimum[openvino,nncf]`         |

Author: @echarlaix, can we say that this is resolved?

Check out the [`examples`](https://github.com/huggingface/optimum-intel/tree/main/examples) directory for more sophisticated usage.

### OpenVINO

> Below are the examples of how to use OpenVINO and its [NNCF](https://docs.openvino.ai/latest/tmo_introduction.html) tool for model optimization and inference.

Reviewer: ```suggestion
Below are the examples of how to use OpenVINO and its [NNCF](https://docs.openvino.ai/latest/tmo_introduction.html) framework for model optimization, quantization, and inference.
```
NNCF framework is doubling framework, but to me tool does not cover it.
IMO it would be good to mention quantization specifically, to prevent confusion about the kind of optimization OpenVINO Runtime does and what NNCF does.

> - OpenVINO inference example:

Reviewer: ```suggestion
#### OpenVINO inference example:
```
IMO it looks nicer if this is a header, and it makes it possible to create a direct link to an example.

pipe_cls = pipeline("text-classification", model=model, tokenizer=tokenizer)
text = "He's a dreadful magician."
outputs = pipe_cls(text)
```

> - Post-training quantizaiton example:

Reviewer: ```suggestion
- Post-training quantization example:
```

)
quantizer.quantize(
    calibration_dataset=calibration_dataset,
    save_directory='nncf_results' # use this folder as model path for OVModelForSequenceClassification
                                  # to do the inference
)
```

- Quantization-aware training example:

```diff
import numpy as np
from datasets import load_dataset, load_metric
from transformers import AutoModelForSequenceClassification, AutoTokenizer, TrainingArguments, default_data_collator
from optimum.intel.openvino.trainer import OVTrainer

model_id = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(model_id)    
tokenizer = AutoTokenizer.from_pretrained(model_id)
dataset = load_dataset("glue", "sst2")
dataset = dataset.map(
    lambda examples: tokenizer(examples["sentence"], padding="max_length", max_length=128), batched=True
)
compute_metrics = lambda p: metric.compute(
    predictions=np.argmax(p.predictions, axis=1), references=p.label_ids
)

-trainer = Trainer(
+trainer = OVTrainer(
    model=model,
    feature="sequence-classification",
    args=TrainingArguments('nncf_results', num_train_epochs=1.0, do_train=True, do_eval=True),

>     train_dataset=train_dataset,

Author: I like the suggestion. It would be helpful for those who does not have GPU but would like to see how the tuning works in practice.

pipe_cls = pipeline("text-classification", model=model, tokenizer=tokenizer)
text = "He's a dreadful magician."
outputs = pipe_cls(text)
```

> - Post-training quantizaiton example:

Reviewer: ```suggestion
#### Post-training quantization example:
```

```python
from functools import partial
from optimum.intel.openvino.quantization import OVQuantizer 
from transformers import AutoTokenizer, AutoModelForSequenceClassification, OVModelForSequenceClassification

model_id = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(model_id)    
tokenizer = AutoTokenizer.from_pretrained(model_id)
def preprocess_fn(examples, tokenizer):
    return tokenizer(

>         examples["sentence"], padding="max_length", max_length=tokenizer.model_max_length, truncation=True

Reviewer: ```suggestion
        examples["sentence"], padding=True, truncation=True, max_length=tokenizer.model_max_length
```
See comment below for QAT example

    )

quantizer = OVQuantizer.from_pretrained(model)
calibration_dataset = quantizer.get_calibration_dataset(
    "glue",
    dataset_config_name="sst2",
    preprocess_function=partial(preprocess_fn, tokenizer=tokenizer),
    num_samples=100,
    dataset_split="train",
    preprocess_batch=True,
)
quantizer.quantize(
    calibration_dataset=calibration_dataset,
    save_directory='nncf_results' # use this folder as and input for OVModelForSequenceClassification
                                  # to do the inference
)

optimized_model = OVModelForSequenceClassification.from_pretrained('nncf_results')
```

> - Quantization-aware training example:

Reviewer: ```suggestion
#### Quantization-aware training example:
```

```diff
import numpy as np
from datasets import load_dataset, load_metric
from transformers import AutoModelForSequenceClassification, AutoTokenizer, TrainingArguments, default_data_collator
from optimum.intel.openvino.trainer import OVTrainer
+from optimum.intel.openvino import OVModelForSequenceClassification

model_id = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(model_id)    
tokenizer = AutoTokenizer.from_pretrained(model_id)
dataset = load_dataset("glue", "sst2")
dataset = dataset.map(

>     lambda examples: tokenizer(examples["sentence"], padding="max_length", max_length=128), batched=True

Reviewer: ```suggestion
    lambda examples: tokenizer(examples["sentence"], padding=True, truncation=True, max_length=tokenizer.model_max_length), batched=True
```
`padding=True` pads to the largest sequence in the batch, which will usually be shorter than max_length, so this speeds up training/inference time by quite a bit. I added truncation=True for the cases where dataset input is longer than model_max_length. 

tokenizer.model_max_length is not defined for every model so it might be safer to set it to a specific value in this README example (and in the PTQ one), to prevent users running into issues if they modify the example with a different model.

Author: I agree and it would be great to have it as generic as possible.

)

> compute_metrics = lambda p: metric.compute(

Reviewer: metric is not defined yet. Easiest is to add `metric = load_metric("accuracy")` but there is a DeprecationWarning that says to use evaluate instead. @echarlaix  should we add evaluate to nncf extra dependencies and use `metric = evaluate.load("accuracy")`?

Reviewer 2: Yes it's a great point

    predictions=np.argmax(p.predictions, axis=1), references=p.label_ids
)

-trainer = Trainer(
+trainer = OVTrainer(
    model=model,
    feature="sequence-classification",
    args=TrainingArguments('nncf_results', num_train_epochs=1.0, do_train=True, do_eval=True),
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    compute_metrics=compute_metrics,
    tokenizer=tokenizer,
    data_collator=default_data_collator,
)
train_result = trainer.train()

> metrics = trainer.evaluate()

Reviewer: ```suggestion
metrics = trainer.evaluate()
print(metrics)
```
I know users can add this themselves but if it takes a while to train, IMO it is nice if the copy/pasteable example shows how well it worked.

pipe_cls = pipeline("text-classification", model=model, tokenizer=tokenizer)
text = "He's a dreadful magician."
outputs = pipe_cls(text)
```

#### Post-training quantization example:
```python
from functools import partial
from optimum.intel.openvino.quantization import OVQuantizer 
from transformers import AutoTokenizer, AutoModelForSequenceClassification, OVModelForSequenceClassification

model_id = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(model_id)    
tokenizer = AutoTokenizer.from_pretrained(model_id)
def preprocess_fn(examples, tokenizer):
    return tokenizer(

>         examples["sentence"], padding=True, truncation=True, max_length=tokenizer.model_max_length

Reviewer 2: ```suggestion
        examples["sentence"], padding=True, truncation=True, max_length=128
```

    )

quantizer = OVQuantizer.from_pretrained(model)
calibration_dataset = quantizer.get_calibration_dataset(
    "glue",
    dataset_config_name="sst2",
    preprocess_function=partial(preprocess_fn, tokenizer=tokenizer),
    num_samples=100,
    dataset_split="train",
    preprocess_batch=True,
)

> quantizer.quantize(
>     calibration_dataset=calibration_dataset,
>     save_directory='nncf_results' # use this folder as and input for OVModelForSequenceClassification
>                                   # to do the inference
> )
> 
> optimized_model = OVModelForSequenceClassification.from_pretrained('nncf_results')

Reviewer 2: ```suggestion
# The directory where the quantized model will be saved
save_dir = "nncf_results"
# Apply static quantization and save the resulting model in the OpenVINO IR format
quantizer.quantize(calibration_dataset=calibration_dataset, save_directory=save_dir)
# Load the quantized model
optimized_model = OVModelForSequenceClassification.from_pretrained(save_dir)
```

```

#### Quantization-aware training example:

```diff
import numpy as np
from datasets import load_dataset, load_metric
from transformers import AutoModelForSequenceClassification, AutoTokenizer, TrainingArguments, default_data_collator

> from optimum.intel.openvino.trainer import OVTrainer

Reviewer 2: ```suggestion
-from transformers import Trainer
+from optimum.intel.openvino.trainer import OVTrainer
```

+from optimum.intel.openvino import OVModelForSequenceClassification

model_id = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(model_id)    
tokenizer = AutoTokenizer.from_pretrained(model_id)
dataset = load_dataset("glue", "sst2")

> dataset = dataset.map(
>     lambda examples: tokenizer(examples["sentence"], padding=True, truncation=True, max_length=tokenizer.model_max_length), batched=True
> )

Reviewer 2: ```suggestion
dataset = dataset.map(
    lambda examples: tokenizer(examples["sentence"], padding=True, truncation=True, max_length=128), batched=True
)
```

compute_metrics = lambda p: metric.compute(
    predictions=np.argmax(p.predictions, axis=1), references=p.label_ids
)

> 

Reviewer 2: ```
# The directory where the quantized model will be saved
save_dir = "nncf_results"
```

-trainer = Trainer(
+trainer = OVTrainer(
    model=model,
    feature="sequence-classification",

>     args=TrainingArguments('nncf_results', num_train_epochs=1.0, do_train=True, do_eval=True),

Reviewer 2: ```suggestion
    args=TrainingArguments(save_dir, num_train_epochs=1.0, do_train=True, do_eval=True),
```

    train_dataset=dataset["train"].select(range(300)),
    eval_dataset=dataset["validation"],
    compute_metrics=compute_metrics,
    tokenizer=tokenizer,
    data_collator=default_data_collator,
)
train_result = trainer.train()
metrics = trainer.evaluate()

> print(metrics)

Reviewer 2: I would remove this print

Author: @helena-intel, asked to add this. I assume it helps to improve UX since we expect that this will be copy-pasted

trainer.save_model()

> +optimized_model = OVModelForSequenceClassification.from_pretrained('nncf_results')

Reviewer 2: ```suggestion
+optimized_model = OVModelForSequenceClassification.from_pretrained(save_dir)
```

```

> You can find more OpenVINO examples in the corresponding Optimum-Intel documentation.

Reviewer 2: ```suggestion
You can find more OpenVINO examples in the corresponding Optimum Intel documentation.
```
